README.md
saber-log [![Build Status](https://travis-ci.org/ecomfe/saber-log.png)](https://travis-ci.org/ecomfe/saber-log)
===

为使用saber框架的知心各行业手机端提供nsclick统计支持。

## Usage

0. 通过`edp`引入模块：

    edp import saber-log

1. 在`src/app.js`开头处添加
    ```javascript
    var log = require('saber-log');
    log.addDefaultLog({
        'fr' : 'wise-zhixin',
        'pvid' : (new Date()).getTime(),
        'pid' : 341
    });
    log.on();
    ```

2. 在`src/app.js`的`loadAction`方法内添加
    ```javascript
    log.addDefaultLog({
        'page' : url,
        'refer' : document.referrer
    });
    log.sendLog({
        'act': 'pv'
    });
    ```
3. 在需要发送log处添加自定义属性，如
    ```html
    <a href="xxx" data-click='{"act":"a_some_act","mod":"content"}' data-log="btn">xxx</a>
    ```

4. 详细nsclick参数标准请参考  
    [NSCLICK上线流程](http://wiki.babel.baidu.com/twiki/bin/view/Ps/OP/NSCLICK%E4%B8%8A%E7%BA%BF%E6%B5%81%E7%A8%8B)  
    [大搜中间页点击日志规范](http://wiki.babel.baidu.com/twiki/pub/Ps/Rank/UbsTopic/Middle_page/%E5%A4%A7%E6%90%9C%E7%B4%A2%E4%B8%AD%E9%97%B4%E9%A1%B5%E7%82%B9%E5%87%BB%E6%97%A5%E5%BF%97%E8%A7%84%E8%8C%83v0.4.pdf)

## API

### .setLogUrl( url )

设置日志发送域名。默认为`http://nsclick.baidu.com/v.gif?`

### .addDefaultLog( options )

设置日志默认参数。这些参数在每次发送时都会被带上，因此常用来设置基本不变的值，例如和应用相关的信息。

options可选值请参考
[大搜中间页点击日志规范](http://wiki.babel.baidu.com/twiki/pub/Ps/Rank/UbsTopic/Middle_page/%E5%A4%A7%E6%90%9C%E7%B4%A2%E4%B8%AD%E9%97%B4%E9%A1%B5%E7%82%B9%E5%87%BB%E6%97%A5%E5%BF%97%E8%A7%84%E8%8C%83v0.4.pdf)  

常用值有：

* `options.fr` `string` 来源，如`ps-zhixin`等
* `options.pid` `string|number` 区分日志id，各产品线参考
[nsclick目前的pid列表](http://wiki.babel.baidu.com/twiki/bin/view/Ps/OP/NSCLICK%E4%B8%8A%E7%BA%BF%E6%B5%81%E7%A8%8B#%E7%9B%AE%E5%89%8D%E7%9A%84pid%E5%88%97%E8%A1%A8%EF%BC%8C%E6%B2%A1%E6%9C%89%E8%A2%AB%E8%AE%B0%E5%BD%95%E4%B8%8D%E8%B4%9F%E8%B4%A3%E7%BB%B4%E6%8A%A4)
* `options.pvid` `string` 记录一次pv的唯一id，一个页面上所有pvid相同，通常取页面载入时的时间戳或由后端提供
* `options.qid` `string` 大搜一次检索唯一标示，用来merge大搜和中间页的点击日志
* `options.page` `string` 当前页面url
* `options.refer` `string` 当前页面访问的referer

### .sendLog( options )

手动发送一条日志。参数`options`会与和`.addDefaultLog`的参数合并，如果相同以`.sendLog`为准。  
常用来发送pv日志。点击日志由属性`data-log`,`data-click`等自行识别发送，不需要手动调用发送。

### .on()

为当前document.body绑定点击(`click`)事件，用以自行识别并发送日志。  
对于普通应用，在新页面载入时调用一次。  
对于单页应用(如`saber`)，全局调用一次即可。  

### .un()

为当前document.body解绑点击(`click`)事件。

## Test

启动测试服务器

    $ node test/server.js

默认端口为`8848`，可以通过参数修改：

    $ node test/server.js 8080

访问`http://localhost:8848/test/runner.html`


===

[![Saber](https://f.cloud.github.com/assets/157338/1485433/aeb5c72a-4714-11e3-87ae-7ef8ae66e605.png)](http://ecomfe.github.io/saber/)


History.md
0.1.0 / 2014-05-04
==================

* 初始化模块，目前支持click事件向nsclick.baidu.com发送日志。后续会增加其他事件(swipe,drag,xxx)


test/ws.js
var fs = require('fs');
var path = require('path');
var PID_FILE = __dirname + '\\.server.pid';

function getServerPID() {
    var pid = 0;

    if (!fs.existsSync(PID_FILE)) {
        return pid;
    }
    
    pid = parseInt(fs.readFileSync(PID_FILE, 'utf-8'), 10);

    return pid;
}

function setServerPID(pid) {
    fs.writeFileSync(PID_FILE, pid, 'utf-8');
}

var commandHandler = {};

commandHandler.start = function () {
    var port = process.argv[3] || 8848;
    if (!getServerPID()) {
        var child = require('child_process').fork(path.resolve(__dirname, 'server'), [port, true]);
        setServerPID(child.pid);
        child.disconnect();
        console.log('server start on ' + port);
    }
};

commandHandler.stop = function () {
    var pid = getServerPID();

    if (pid) {
        fs.unlinkSync(PID_FILE);
        process.kill(pid, 'SIGKILL');
        console.log('server stop');
    }
};

var cmd = process.argv[2];

var handler = commandHandler[cmd];

if (handler) {
    handler();
}
else {
    console.log(cmd + ' command not found');
}

process.exit(0);


test/server.js
/**
 * @file 测试服务器
 * @author treelite(c.xinle@gmail.com)
 * @author wangyisheng(wangyisheng@outlook.com)
 */

var MIME_TYPES = {
    json : 'application/json',
    js: 'text/javascript',
    html: 'text/html'
};

var IS_SILENT = false;

var fs = require('fs');
var LOG_FILE = __dirname + '\\.saber.log';
if (fs.existsSync(LOG_FILE)) {
    fs.unlinkSync(LOG_FILE);
}

function log(msg) {
    if (!IS_SILENT) {
        console.log('[Server]' + msg);
    }
}

function extend(target, source) {
    for (var key in source) {
        if (source.hasOwnProperty(key)) {
            target[key] = source[key];
        }
    }

    return target;
}

function createResponse(response, data) {
    if (data.headers) {
        for (var key in data.headers) {
            if (data.headers.hasOwnProperty(key)) {
                response.setHeader(key, data.headers[key]);
            }
        }
    }
    if (!response.getHeader('Content-Type')) {
        response.setHeader('Content-Type', 'text/html');
    }
    response.statusCode = data.status || 200;
    response.end(data.content);
}

/**
 * 判断是否是静态文件
 *
 * @inner
 * @param {string} pathname
 * @return {boolean}
 */
function isStatic(pathname) {
    if (pathname.charAt(0) == '/') {
        pathname = pathname.substring(1);
    }
    var path = require('path');
    var fs = require('fs');
    var file = path.resolve(process.cwd(), pathname);

    return fs.existsSync(file);
}

/**
 * 获取静态文件
 *
 * @inner
 * @param {string} pathname
 * @return {Object}
 */
function getStatic(pathname) {
    if (pathname.charAt(0) == '/') {
        pathname = pathname.substring(1);
    }
    var path = require('path');
    var fs = require('fs');
    var file = path.resolve(process.cwd(), pathname);

    return {
        content: fs.readFileSync(file),
        mimetype: MIME_TYPES[path.extname(file).substring(1)] || 'text/html'
    };
}

/**
 * 解析Form-Data数据
 * 简单方式处理
 *
 * @inner
 * @param {string} str 请求参数
 */
function parseFormData(str) {
    var res = {};
    var regexp = new RegExp('Content-Disposition:\\s+form-data;\\s+name="([^"]+)"', 'g');

    var finded;
    var i;
    var dataToken;
    var endToken;
    var token;
    var name;
    while(finded = regexp.exec(str)) {
        name = RegExp.$1;
        dataToken = [];
        endToken = [];
        i = regexp.lastIndex;
        while (i < str.length 
            && (str.charAt(i) == '\r' || str.charAt(i) == '\n')
        ) {
            i++;
        }
        while (i < str.length) {
            token = str.charAt(i);
            dataToken.push(token);
            if (token == '-') {
                endToken.push(token);
            }
            else {
                endToken = [];
            }
            if (endToken.length >= 6) {
                dataToken.splice(
                    dataToken.length - endToken.length, 
                    endToken.length
                );
                dataToken = dataToken.join('').trim();
                if (res[name]) {
                    res[name] = [res[name]];
                    res[name].push(dataToken);
                }
                else {
                    res[name] = dataToken;
                }
                break;
            }
            i++;
        }
    }
    return res;
}

/**
 * 获取POST数据
 * 按文本解析，不支持file
 *
 * @inner
 * @param {Object} request
 * @param {Function(Object)} callback
 */
function getPostData(request, callback) {
    var data = [];

    request.on('data', function (chunk) {
        data.push(chunk);
    });

    request.on('end', function () {
        data = data.join('');
        if (data.indexOf('Content-Disposition: form-data;') >= 0) {
            data = parseFormData(data);
        }
        else {
            data = require('querystring').parse(data);
        }
        callback(data);
    });
}

var actionList = {};

actionList['/v.gif'] = function (request, response) {
    // /v.gif?a=1&b=2
    fs.appendFileSync(LOG_FILE, request.url + '\n');
    createResponse(response, {
        status: 200
    })
};

actionList['/getLog'] = function (request, response) {
    if (!fs.existsSync(LOG_FILE)) {
        createResponse(response, {
            status: 500
        });
    } else {
        var queryString = require('url').parse(request.url, true).query;
        var type = queryString.type;
        var arr = fs.readFileSync(LOG_FILE).toString('utf-8').trim().split('\n');
        var pvLog, clickLog;
        // 读最后两条
        if (arr[arr.length - 1].indexOf('act=pv') != -1) {
            pvLog = arr[arr.length - 1];
            clickLog = arr[arr.length - 2];
        } else {
            pvLog = arr[arr.length - 2];
            clickLog = arr[arr.length - 1];
        }
        createResponse(response, {
            content: type == 'pv' ? pvLog : clickLog,
            status: 200
        })
    }
};

var http = require('http');

var port = process.argv[2] || 8848;

IS_SILENT = process.argv[3] == 'true';

var server = http.createServer();

var urlMgr = require('url');
server.on('request', function (request, response) {
    var url = urlMgr.parse(request.url);
    var pathname = url.pathname;
    log('[Request]' + pathname);
    var action = actionList[pathname];
    if (action) {
        action(request, response);
    }
    else if (isStatic(pathname)) {
        var data = getStatic(pathname);

        response.statusCode = 200;
        response.setHeader('Content-Type', data.mimetype);
        response.end(data.content);
    }
    else {
        response.statusCode = 404;
        response.setHeader('Content-Type', 'text/html');
        response.end('not found');
    }
});

server.listen(port);
log('server start at ' + port + '...');


test/main.spec.js
/**
 * @file log.spec ~ 2014-05-04 14:55:31
 * @author easonyq(wangyisheng@outlook.com)
 * @description
 * saber-log test
 */

define(function() {
    var logger = require('saber-log');
    var ajax = require('saber-ajax');

    // 设置默认参数
    logger.setLogUrl('http://localhost:8848/v.gif?');
    var pvid = (new Date()).getTime().toString();
    logger.addDefaultLog({
        'fr' : 'saber-log-test',
        'pvid' : pvid,
        'pid' : '0',
        'page' : 'saber-log-test-page'
    });

    // 绑定文档
    logger.on();

    // 发送pv日志
    logger.sendLog({
        'act': 'pv'
    });

    // 代码触发click事件，发送click日志
    var a = document.getElementById('testLink');
    if (a.click && (typeof a.click == 'function')) {
        a.click();
    } else {
        var evt = document.createEvent("MouseEvents");
        evt.initEvent("click",true,true);   
        a.dispatchEvent(evt);
    }

    // 把log从url格式转化为obj
    function processLog(log) {
        var param = log.substring(log.indexOf('?') + 1, log.length);
        var arr = param.split('&');
        var result = {};
        for (var index in arr) {
            var tmp = arr[index].split('=');
            if (tmp[1].length != 0) {
                result[tmp[0]] = decodeURIComponent(tmp[1]);
            }
        }
        return result;
    }

    describe('pvLog', function() {
        var log;
        beforeEach(function(done) {
            ajax.get('/getLog?type=pv').then(function(data) {
                log = processLog(data);
                done();
            });
        });

        it('correctness', function() {
            expect(log).toEqual({
                'act' : 'pv',
                'fr' : 'saber-log-test',
                'page' : 'saber-log-test-page',
                'pid' : '0',
                'pvid' : pvid
            });
        });
    });

    describe('clickLog', function() {
        var log;
        beforeEach(function(done) {
            ajax.get('/getLog?type=click').then(function(data) {
                log = processLog(data);
                done();
            });
        });

        it('correctness', function() {
            // expect(true).toBe(true);
            expect(log).toEqual({
                'act' : 'b_test',
                'fr' : 'saber-log-test',
                'page' : 'saber-log-test-page',
                'pid' : '0',
                'pvid' : pvid,
                'mod' : 'test',
                'xpath' : 'ul-li4-a(btn)',
                'url' : 'http://localhost:8848/test/runner.html#',
                'txt' : 'Click Here!'
            });
        });
    });
}) 


src/main.js
/**
 * @file main ~ 2014-05-04 11:51:33
 * @author easonyq(wangyisheng@outlook.com)
 * @description
 * saber-log主文件
 */
define(function() {

    var lang = require('saber-lang');

    /**
     * 是否已经初始化
     * @type {boolean}
     */
    var isInited = false;
    /**
     * 发送日志的目的地
     * @type {string}
     */
    var LOG_URL = 'http://nsclick.baidu.com/v.gif?';

    /**
     * 每次发会附加的参数
     * @type {Object}
     * @private
     */
    var _defaultLog = { }; 

    /**
     * 记录本次发送的日志数据
     * @type {Object}
     * @private
     */
    var _logData = { };

    /**
     * 本次事件的target
     * @type {Node|undefined}
     * @private
     */
    var _target;

    /**
     * 事件
     * @type {Event}
     * @private
     */
    var _event;

    /**
     * 发送日志
     * @param {Object} params 需要发送的参数
     */
    function sendLog(params) {
        var parameters = {};

        lang.extend(parameters, _defaultLog);
        lang.extend(parameters, params);

        var tmpUrl = [];
        for (var key in parameters) {
            tmpUrl.push(key + '=' + encodeURIComponent(parameters[key]));
        }
        var url = tmpUrl.join('&');
        var img = new Image();
        img.onload = img.onerror = img.onabort = function() {
            img.onload = img.onerror = img.onabort = null;
            img = null;
        };

        img.src = LOG_URL + url;
    }

    function addLogData(data) {
        if (data) {
            lang.extend(_logData, data);
        }
    }

    function clearLogData() {
        _logData = { };
    }

    function collectLogData() {
        clearLogData();

        var target = _target;
        /**
         * 本次统计的数据
         */
        var logData = {};
        /**
         * url参数
         */
        var url;
        /**
         * 点击或操作对象
         */
        var item;
        /**
         * 表示此次点击target的类型
         */
        var type;
        /**
         * 匹配data-log值的
         */
        var typeReg = /\b(title|link|img|btn|input|others)\b/i;
        /**
         * 到新页面a[\w+]还是在本页面b[\w+]
         */
        var act = 'b';
        /**
         * 记录xpath
         */
        var path = [];
        var i = 0;
        while (target && target !== document.body) {
            if (target.getAttribute('data-nolog') === '1') {
                // 如果遇到标签或者父标签上有data-nolog属性，则不发送日志
                return false;
            }
            // 添加标签上的data-click数据
            logStr = target.getAttribute('data-click');
            if (logStr) {
                try {
                    logData = lang.extend({}, JSON.parse(logStr), logData);
                } 
                catch(e) {}
            }
            if (target.href) {
                url = target.href;
                type = 'link';
                act = target.getAttribute('target') === '_blank' ? 'a' : act;
            }
            // 如果target上有data-log属性，并且data-log=button, link等，则需要发送统计
            var dataLogAttr = target.getAttribute('data-log')
            if (typeReg.test(dataLogAttr)) {
                type = RegExp.$1.toLowerCase();
            }
            // 如果是个链接，则添加链接 /^a|img|input|button|select|datalist|textarea$/
            var count = 1;
            if (target.previousSibling) {
                var sibling = target.previousSibling;
                do {
                    if (sibling.nodeType === 1
                        && sibling.tagName === target.tagName) {
                       count++;
                   }
                   sibling = sibling.previousSibling;
                }
                while (sibling);
            }
            path[i++] = target.tagName + (count > 1 ? count : '');
            if (target.getAttribute('data-id')) {
                item = target.getAttribute('data-id');
            }
            target = target.parentNode;
        }
        path.reverse();

        var tag = _target.tagName.toLowerCase();
        if (!type && /^(a|img|input|button|select|datalist|textarea)$/.test(tag)) {
            type = {a: 'link'}[tag] || 'input';
            url = _target.href || _target.src || url;
        }
        if (!type) {
            return false;
        }
        var title;
        if (tag == 'a') {
            title = _target.innerHTML;
        } else if (type === 'input') {
            // 如果是表单元素
            if(/input|textarea/.test(tag)) {
                title = _target.value;
                if (_target.type && _target.type.toLowerCase() === 'password') {
                    title = '';
                }
            } else if (/select|datalist/.test(tag)) {
                if (_target.children.length > 0) {
                    var index = _target.selectedIndex || 0;
                    title = _target.children[index > -1 ? index : 0].innerHTML;
                }
            } else {
                title = _target.innerText || _target.value || '';
            }
        } else {
            // 如果是图片，先取其title
            if (tag === 'img') {
                title = _target.title;
            }

            // title为空，遍历父节点
            if (!title) {
                var el = _target;
                while (i > 0) {
                    i--;
                    if (/^a\d*\b/i.test(path[i])) {
                        url = el.href;
                        title = el.innerText;
                        break;
                    } else {
                        if (typeReg.test(el.getAttribute('data-log'))) {
                            title = el.innerText;
                            break;
                        }
                        el = el.parentNode;
                    }
                }
            }
        }
        title && (title = title.trim());
        /**
         * xpath值
         */
        var xpath = path.join('-').toLowerCase() + '(' + type + ')';
        lang.extend(logData, {
            'txt': title || '',
            'xpath': xpath,
            'url': url || '',
            'act': logData['act'] || act
        });
        if (item) {
            logData['item'] = logData['item'] || item;
        }
        addLogData(logData);
        return true;
    }

    /**
     * 事件的handler
     */
    function _eventHandler(event) {
        _event = event;
        _target = event.target;

        // 如果collectLogData返回false，则不发送日志
        if (collectLogData() === false) {
            return;
        }
        sendLog(_logData);
    }

    function bind() {
        if (isInited) {
            return;
        }
        isInited = true;
        document.body.addEventListener('click', _eventHandler);
    }

    function unbind() {
        document.body.removeEventListener('click', _eventHandler);
    }

    var exports = {
        /**
         * 发送日志
         */
        sendLog: sendLog,
        /**
         * 绑定事件
         */
        on: bind,
        /**
         * 解绑事件
         */
        un: unbind,
        /**
         * 添加默认参数
         * @param {Object} data
         */
        addDefaultLog: function(data) {
            lang.extend(_defaultLog, data);
        },
        /**
         * 设置发送日志域名
         * @param {string} url
         */
        setLogUrl: function(url) {
            LOG_URL = url;
        }
    };

    return exports;
});